contracts/auctions/FungibleAuction.sol

```diff
@@ -0,0 +1,172 @@
+// TODO: Use interface. The current version of @openZeppelin/contracts lib has none that fits.
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+// TODO: Switch to proper interface?
+interface IAuctionHouse {
+    function closeFungible(
+        address buyTokenAddress,
+        address sellTokenAddress,
+        uint256 tokensSold,
+        uint256 minPrice
+    ) external;
+}
+
+/// @title Auction for fungible tokens
+contract FungibleAuction {
+    using SafeMath for uint256;
+
+    /// @dev Only set if auction was created via the auction house factory. Used for callbacks.
+    IAuctionHouse public auctionHouse;
+
+    /// @notice Address of the contract governing the token that is being auctioned.
+    ERC20 public immutable sellTokenAddress;
+
+    /// @notice Address of the contract governing the token used to buy in this auction.
+    ERC20 public immutable buyTokenAddress;
+
+    /// @notice Creator of the auction and address auctioning their tokens.
+    address public immutable auctioneer;
+
+    /// @notice Parameters of this specific auction.
+    uint256 public immutable initialAmount;
+    uint256 public immutable startTimestamp;
+    uint256 public immutable duration;
+    uint256 public immutable maxPrice;
+    uint256 public immutable minPrice;
+
+    /// @notice The last paid price, for Dutch Auctions this is also the lowest price.
+    uint256 public lowestSellPrice;
+
+    constructor(
+        ERC20 sellTokenAddress_,
+        ERC20 buyTokenAddress_,
+        uint256 initialAmount_,
+        address auctioneer_,
+        uint256 startTimestamp_,
+        uint256 duration_,
+        uint256 maxPrice_,
+        uint256 minPrice_,
+        bool isAuctionHouse
+    ) {
+        require(maxPrice_ >= minPrice_, "Auction: Price must go down");
+        require((startTimestamp_ + duration_) >= block.timestamp, "Auction: Cannot end in the past");
+
+        if (isAuctionHouse) {
+            auctionHouse = IAuctionHouse(msg.sender);
+        }
+
+        initialAmount = initialAmount_;
+        auctioneer = auctioneer_;
+        sellTokenAddress = sellTokenAddress_;
+        buyTokenAddress = buyTokenAddress_;
+        startTimestamp = startTimestamp_;
+        duration = duration_;
+        maxPrice = maxPrice_;
+        minPrice = minPrice_;
+    }
+
+    /// @dev Get price per token.
+    function getPrice() public view returns (uint256) {
+        if (block.timestamp <= startTimestamp) {
+            return maxPrice;
+        }
+
+        // Guard against timestamps *very* far into the future and/or long durations.
+        // TODO: Intended? Should an item that can no longer be sold return `minPrice`?
+        if (block.timestamp > startTimestamp.add(duration)) {
+            return minPrice;
+        }
+
+        // Safe subtraction. Already guarded in the constructor.
+        uint256 priceRange = maxPrice - minPrice;
+
+        // Safe subtraction. Already guarded via line 82.
```

Comment:
Comments with reference to exact lines are a bit fragile

Reply:
Agreed. Will change!